### ISSUE TITLE
Restrict regex to a subset of XML regex specification

### ISSUE THREAD

**Issue by rgwilton** [netmod-wg/yang-next#21]
The choice in YANG to use XML regex has effectively meant that there is only a single library implementation of regex parsing.

In most cases, the pattern statements only use the basic subset of the XML regex, and normally these pattern statements would validate against most standard regex engines with only a minimal amount of changes (it might be necessary to add anchors at the start and end of the line.

Hence this proposal for the next version of YANG is to restrict the supported regex to a subset of the XML regex language.  Perhaps something along the following:
 -  disallow subtractions in character classes (e.g. [A-Z-[P-R]])
 - disallow matching against various control characters (bell, backslash, escape, form feed, etc)
 - disallow against XML shorthands (\i, \c)
-  disallow against unicode blocks (e.g. \p{IsArrows})? 
 - advise that the whole string is always matched, so start/end anchors are not necessary.
 - check that line end matching is consistent
 - clarify that \d and \w are Unicode matches (rather than ASCII) and implementations must translate them to equivalent character class matches if required.

**Comment by schoenw** [comment]
I am against creating yet another flavor of regular expressions.

**Comment by kwatsen** [comment]
idea is that this would be a subset of XML regex that would so happen to be a subset of POSIX regex...

**Comment by schoenw** [comment]
It is still a new flavour of a regex and I am against creating new flavors of regex.

**Comment by mbj4668** [comment]
I agree with Juergen.  Also, even if the subset happens to be legal POSIX, it doesn't mean the same thing (due to different anchoring rules).
Also, defining our own regexp flavor will not make implementations simpler, on the contrary, with this I can't use _any_ 3rd party regexp library.

**Comment by rgwilton** [comment]
I believe that the POSIX start/end line anchors are a minor difference and trivial to mitigate.

The issue that I have is that some (perhaps many) implementations put the YANG pattern statements directly into whatever the default language regex engine is.  95% of the time that will work fine (e.g. if they only use the subset I propose above), but if they don't then there is going to be the odd interop problem.

Since, what I'm proposing is a subset of XML regex, then it would be guaranteed to at least work in the same XML RE libraries being used today, and the intention is that it should work in all other major regex libraries as well.

So, I see that the only difference is on the tooling that validates that a YANG module is valid YANG since this would need to be coded to check the stricter pattern statement rules to be classified as valid YANG.

**Comment by schoenw** [comment]
We meanwhile have RFC 9485, another regular expression format. RFC 9485 claims that their I-Regexp is a proper subset of XSD regex and several other regex formats. If true, adopting the RFC 9485 may be a solution.

**Comment by rovarga** [comment]
I agree with @schoenw based on RFC9485 description:
```
This is essentially an XSD regexp without:
character class subtraction, multi-character escapes such as \s, \S, and \w, and Unicode blocks.
```
and [XSD interop note](https://www.rfc-editor.org/rfc/rfc9485#name-xsd-regexps).

As in implementor, [this has been a thorn in my side](https://lf-opendaylight.atlassian.net/browse/YANGTOOLS-887) -- especially the three areas excluded are concerned.

While switching to I-Regexp will not ease my burden as a 1.0 and 1.1 implementation, it certainly will help going forward. If we can have any (and preferrably all) of:
- a `yang-version 1.2` which mandates I-Regexp
- a rfc8407bis mandating the use I-Regexp
- rfc6991bis adopting I-Regexp (I have not checked the current state)

I would very much appreciate it. Having the latter two would go a long way in ensuring implementations actually do The Right Thing(tm), which is most likely not the case right now.

**Comment by kwatsen** [comment]
Th should instead use RFC 9485.  The following is from Gemini AI:

[RFC 9485](https://datatracker.ietf.org/doc/html/rfc9485) defines I-Regexp (Interoperable Regular Expression), a lightweight and restricted regular expression format created in 2023 by C. Bormann and T. Bray. 

Core Features
  - Boolean Matching Only: It tests whether a pattern matches an entire string, returning true or false without capturing substrings.
  - Implicit Anchoring: Patterns are automatically anchored, behaving as if they start with ^ and end with $.
  - No Advanced Features: It completely omits capture groups, lookarounds, and backreferences to ensure predictable performance.
  - Unicode Support: It supports the full repertoire of Unicode scalar values and properties in character classes.
  - Linear-Time Safety: It avoids complex backtracking bugs and security risks like ReDoS (Regular Expression Denial of Service). 

Motivation
  - The format was developed for the [IETF JSONPath Working Group](https://datatracker.ietf.org/doc/rfc9485/) to allow safe, consistent filter matching across different programming languages and regex engines. Read the official specification on the IETF Datatracker.
